packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts

```diff
@@ -0,0 +1,500 @@
+    from: { columnModelIndex: number; jsonSubColumnModelIndex?: number }
+    to: { columnModelIndex: number; jsonSubColumnModelIndex?: number }
+  },
+  prevState: ColumnModelFormData[],
+) {
+  const { from, to } = action
+  const arrayToReorder =
+    from.jsonSubColumnModelIndex !== undefined
+      ? prevState[from.columnModelIndex].jsonSubColumns!
+      : prevState
+
+  const fromIndex =
+    from.jsonSubColumnModelIndex !== undefined
+      ? from.jsonSubColumnModelIndex
+      : from.columnModelIndex
+
+  const toIndex =
+    to.jsonSubColumnModelIndex !== undefined
+      ? to.jsonSubColumnModelIndex
+      : to.columnModelIndex
+
+  // Move the element
+  const newArray = moveElementInArray(arrayToReorder, fromIndex, toIndex)
+
+  if (from.jsonSubColumnModelIndex !== undefined) {
+    // If we moved a jsonSubColumn, update state to use the new, reordered array
+    prevState[from.columnModelIndex].jsonSubColumns =
+      newArray as JsonSubColumnModelFormData[]
+    return prevState
+  } else {
+    // Otherwise, return the new, reordered array of column models
+    return newArray
+  }
+}
+
+export function reducer(
+  prevState: ColumnModelFormData[],
+  action: TableColumnSchemaFormReducerAction,
+) {
+  switch (action.type) {
+    case 'setValue':
+      prevState = action.value
+      break
+    case 'toggleSelectAll': {
+      prevState = toggleSelectAll(prevState)
+      break
+    }
+    case 'setColumnModelValue': {
+      setColumnModelValue(action, prevState)
+      break
+    }
+    case 'changeColumnModelType': {
+      changeColumnModelType(action, prevState)
+      break
+    }
+
+    case 'appendColumn':
+      prevState.push(getDefaultColumnModelFormData())
+      break
+    case 'appendJsonSubColumn': {
+      const { columnModelIndex } = action
+      prevState[columnModelIndex] = {
+        ...prevState[columnModelIndex],
+        jsonSubColumns: [
+          ...(prevState[columnModelIndex].jsonSubColumns ?? []),
+          getDefaultJsonSubColumnFormData(),
+        ],
+      }
+      break
+    }
+    case 'toggleSelect': {
+      toggleSelect(action, prevState)
+      break
+    }
+    case 'delete':
+      prevState = deleteColumnModel(prevState)
+      break
+    case 'move': {
+      prevState = moveColumnModel(action, prevState)
+      break
+    }
+    case 'moveUp':
+      prevState = moveSelectedItemsUp(prevState, cm => cm.isSelected)
+      prevState.forEach(cm => {
+        if (cm.jsonSubColumns) {
+          cm.jsonSubColumns = moveSelectedItemsUp(
+            cm.jsonSubColumns,
+            jsc => jsc.isSelected,
+          )
+        }
+      })
+      break
+    case 'moveDown':
+      prevState = moveSelectedItemsDown(prevState, cm => cm.isSelected)
+      prevState.forEach(cm => {
+        if (cm.jsonSubColumns) {
+          cm.jsonSubColumns = moveSelectedItemsDown(
+            cm.jsonSubColumns,
+            jsc => jsc.isSelected,
+          )
+        }
+      })
+      break
```

Comment:
`moveUp`/`moveDown` would handle moving selected items (i.e. the current paradigm)